### ISSUE TITLE
NixOS Calamares installer

### ISSUE THREAD

**Issue by jonringer** [NixOS/nixpkgs#100475]
I learned a few weeks ago about [Calamares](https://github.com/calamares/calamares) . A nice distro-independent graphical installer for linux. I think it would be really beneficial for people to avoid the "most difficult" part of getting NixOS on bare metal.

I haven't done any feasibility, just curious if people feel strongly for or against this.

**Comment by FRidh** [comment]
Graphical installer topic also suggesting Calamares https://github.com/NixOS/nixpkgs/issues/21662.

**Comment by jonringer** [comment]
haha, I just stumbled upon that thread. I saw the thread update with your post above :)

**Comment by jonringer** [comment]
crazy how that was "only" 3 1/2 years ago, and there's been 80,000 new issues/PRs since then O.o

**Comment by AndersonTorres** [comment]
Would be OK to close it in favor of #21662?

**Comment by jonringer** [comment]
I think the two are related, but not the same.

The other is proposing a graphical installer in general, and I'm proposing one specifically.

Perhaps discourse would be a better forum, as there's likely to be a narrative there, and the discoverability should be better as well.

**Comment by lboklin** [comment]
Related discussion on Discourse: https://discourse.nixos.org/t/user-friendly-nixos-distro/

**Comment by davidak** [comment]
People started to work on this at NixCon 2019. @manveru made the last commit about a year ago in https://github.com/Lucus16/calamares/tree/os-modules. So there would be the place to continue.

https://discourse.nixos.org/t/what-are-your-goals-for-20-03/4773/7?u=davidak

@manveru @Lucus16 do you want to continue the work? Can you get it in a clean state with comments etc. so someone else can continue the work?

It would be great to have a basic version ASAP. Just like the manual with ext4 and standard configuration.nix. Then we can build on that. Don't announce in release notes until you can select a DE.

**Comment by Lucus16** [comment]
If anyone is willing to pick this up, just email or @ me and I'm willing to
guide you through what we wrote at least. I might help out too after that.

**Comment by manveru** [comment]
I also wouldn't mind helping out on this again, it was mostly held up because I tried to include a nice way to select different profiles for DEs and common use-cases, but the Calamares UI was just not able to provide that. Haven't checked it in a while, so hopefully things are better now, or someone with C++ skills could jump in and provide the needed functionality.

**Comment by cwhy** [comment]
Elementary os have released another installer. Is it worth a look? https://medium.com/elementaryos/meet-the-upcoming-installer-a499341218f5

**Comment by davidak** [comment]
@cwhy the frontend has no release yet. maybe with their next os release https://github.com/elementary/installer/releases

It was suggested before when it was announced: https://github.com/NixOS/nixpkgs/issues/21662#issuecomment-377938406

I think it would be great in combination with Pantheon desktop, but for a general purpose installer, Calamares might be better.

**Comment by aanderse** [comment]
I assume it is written in `vala` which is infinitely more obscure than `c++` so probably not a good idea.

**Comment by samuela** [comment]
It's not GUI, but I created https://github.com/samuela/nixos-up for exactly this purpose. IMHO as long as it doesn't make the user think too much, I think a CLI installer is fine. With that in mind nixos-up strives to get you into a working install as quickly as possible, with the assumption that the user can modify settings to their personal liking from there.

Some ways nixos-up could be improved:
* Support resizing existing partitions such that it's not a "wipe the disk" kind of install.
* Integrate more smoothly into the ISO experience so the user doesn't have to write `sudo nix-shell <long-url...>`.

**Comment by jonringer** [comment]
I was poking around for inspiration, and found that there's already a nixos module https://github.com/calamares/calamares-extensions/tree/calamares/modules/os-nixos

cc @adriaandegroot for inclusion

**Comment by manveru** [comment]
That module was meant as an example for us when we started working on the installer, it's not actually doing anything useful.

**Comment by jonringer** [comment]
Thought it was interesting :)

**Comment by adriaandegroot** [comment]
> That module was meant as an example for us when we started working on the installer, it's not actually doing anything useful.

Yep. It's a proof-of-not-much-concept; the idea behind it was basically "you could use Calamares as the UI to collect configuration things (language, keyboard, user, ..) and do the basics like partitioning and EFI setup and writing an initial filesystem (e.g. from a squashfs), and **then** the *os-nixos* module could take over for all the special configurations needed.

I don't know a thing about NixOS beyond the "there's hashes in paths where I wouldn't expect them", so any progress there depends on questions / issues / PRs from NixOS folks.

**Comment by vlinkz** [comment]
I used @adriaandegroot module as a base for a pair of modules I made in https://github.com/vlinkz/calamares-nixos-extensions. The first module [nixos-generate-config](https://github.com/vlinkz/calamares-nixos-extensions/tree/calamares/modules/nixos-generate-config) creates an initial filesystem and the second [nixos-customize-config](https://github.com/vlinkz/calamares-nixos-extensions/tree/calamares/modules/nixos-customize-config) adds options from the locale, keyboard, and users modules. Also to fix some broken modules I opened [#151508](https://github.com/NixOS/nixpkgs/pull/151508) Still haven't worked on branding at all, but tested a few basic installs with the modules and they seemed to work perfectly as expected. Haven't looked into adding new package options, but I think it shouldn't be too difficult given calamares' packaging module. Any thoughts on my implementation?
